File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    answer = []
    i = 1
    x,y=0,0
    direction = 0
    snail = [[0] * i for i in range(1, n+1)]
    
    while n*n :
        snail[y][x] = i
        i +=1
        if direction == 0 : #down
            if y+1<n and snail[y+1][x] ==0:
                y+=1
            elif x+1<n and snail[y][x+1] == 0:
                x+=1
                direction = 1
            else :
                break
        elif direction == 1: # right
            if x+1<n and snail[y][x+1] == 0:
                x+=1
            elif y-1>0 and x-1>0 and snail[y-1][x-1] == 0:
                y-=1
                x-=1
                direction =2
            else :
                break
        elif direction ==2 : # top
            if y-1>0 and x-1>0 and snail[y-1][x-1] == 0:
                x-=1
                y-=1
            elif y+1<n and snail[y+1][x] ==0:
                y+=1
                direction=0
            else :
                break
        
    answer = sum(snail,[])

    return answer

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
    var answer = [];
    let size = 0
    let control = 0
    let [x,y] = [0,0]
    
    const snail = [... new Array(n)]
    .map((_,i)=>{
        size += i+1;
        return [... new Array(i+1)].fill(0);
    })
    
    for (let i =1; i< size+1; i++){
        snail[y][x] = i;
        if (control === 0){ // down
            if (y+1<snail.length && snail[y+1][x] === 0) y++;
            else {
                control = 1;
                x++;
            }
        } else if (control === 1){  // right
            if (x<snail[y].length && snail[y][x+1] === 0) x++;
            else {
                control = 2;
                y--;
                x--;
            }
        } else{ // up
            if (y-1>0 &&snail[y-1][x-1] === 0){
                y--;
                x--;
            }
            else{
                control = 0;
                y++;
            }
        } 
    }
    //  모든 하위 배열 요소를 지정한 깊이까지 재귀적으로 이어붙인 새로운 배열을 생성
    answer = snail.flat();  
    return answer;
}